### PR TITLE
Add WithLogger option to supply a fxevent.Logger

### PR DESCRIPTION
**Pull request by shirchen**
This is a follow up to #717. This change adds a `WithLogger` option that
accepts a constructor that returns an `fxevent.Logger`. Unlike
`fx.Printer`, this constructor can have dependencies on other types
provided to the Fx graph. If supplied, we'll use this logger to log
events rather than our default behavior.

To make this work, we do the following:

- If `WithLogger` was not supplied, we use the old behavior of logging
  to stderr or the user-supplied `fx.Printer`.
- If `WithLogger` was supplied, we buffer log messages in-memory until
  we can build the user-supplied logger. If we successfully build the
  logger, we flush the buffered messages to it, and log to it from that
  point onwards. If we were unable to build the logger, we flush the
  buffered messages to stderr, or the user-supplied `fx.Printer` and
  fail application startup.

Further, to match our old behavior more accurately, this change includes
a `ConsoleLogger` implementation of `fxevent.Logger` that behaves almost
exactly like our logger before #717. This will make it easy to roll this
feature out without affecting users negatively, and also give us the
option of selecting between readable and non-readable log outputs based
on runtime environment (production vs development, for example).

Ref GO-77

### CODE REVIEW

File: app.go
Comment (r-hang) [inline app.go:463]: As I read it, the value of logBuffer is to store log entries to write before the constructed structured logger is initialized so that those entries can be logged with the structure logger as opposed to another default logger for a more consistent experience.

Is this logger field within logBuffer required? Having a buffer with a logger allows someone to log with both the 'app.log' when its setup as well as the logBuffer once' it's connected, where ideally there should only be a single choice?

The alternative approach would be to replace Connect with a `func (l *logBuffer) Sync(logger fxlog.Logger)`
that only flushes and does not set a logger. Once the structured logger is available, the buffer shouldn't be used anymore? Is there a strong argument against going the other way and even closing off the logBuffer from further writes once its Synced?

File: app.go
Comment (r-hang) [inline app.go:479]: Use l.logger.Log(e) since it's already set.

File: app.go
Comment (r-hang) [inline app.go:525]: If app returns here, will anything written to app.logBuffer ever get flushed out of the buffer?

File: app.go
Comment (abhinav) [inline app.go:461]: readability nit: this file has gotten pretty large. consider moving logBuffer to a sibling file (log.go?) so that we can also independently unit test it.

File: app.go
Comment (abhinav) [inline app.go:491]: nit: just logBuffer as the name should suffice. Although in this case, we don't actually need the variable.

```suggestion
	app := &App{
		logBuffer:    new(logBuffer),
```

File: app.go
Comment (abhinav) [inline app.go:512]: I'm not sure but I think this check should be after `opt.apply` but before `app.provide`? By this point in time, we've already provided a bunch of things to the container and logged about them. We also probably need an `else` clause to use the default logger when this is unspecified?

So what we need here is probably something like:

```
for _, opt := range opts {
  opt.apply(app)
}

var connectLogger interface{}
if app.fxLogConstructor != nil {
    // There's a user provided constructor. Log to buffer until we connect.
    app.log = app.logBuffer
    connectLogger = func(log fxlog.Logger) {
        app.logBuffer.Connect(log)
    }
} else {
    app.log = // put the default logger here
}

// ... app.provide(p)

if connectLogger != nil {
    app.container.Invoke(connectLogger)
}
```

The net result of this will be:

- Existing logging code remains unchanged and always logs to app.log
- app.log is a buffer if the user provided a logger. Otherwise it's not.
- If the user provided a logger, then the first thing we do after all provides is connect the logger

---

On top of this, we can actually get away with not having logBuffer as a field on app:

```
if app.fxLogConstructor != nil {
  var buffer logBuffer
  app.log = &buffer
  connectLogger = buffer.Connect // func(fxlog.Logger)
}
```

I slightly prefer this so that we don't accidentally use the logBuffer directly ever.

---

Finally, and this is optional, but the the Invoke can instead be this:

```
func(log fxlog.Logger) {
  buffer.FlushTo(log)
  app.log = log  // replace buffer with real logger
}
```

This will remove the buffer completely from the equation once things are ready and remove the if-else from the buffer. The buffer would just record things in memory until the replacement is available at which point we flush to replacement and replace.

File: app.go
Comment (abhinav) [inline app.go:537]: We probably want the logBuffer.Connect as suggested above so that it happens after all the other Provides.

That way, the log constructor can actually have a dependency on the fx.Lifecycle. (Which makes sense in case the logger has some kind of Shutdown functionality.)

File: app.go
Comment (abhinav) [inline app.go:706]: This should be reverted so that we're always writing to app.log, and app.log just points to something else.

File: app.go
Comment (abhinav) [reply app.go:525]: This is valid. If the Invoke of the fxlog.Logger fails, that means we failed to construct the user provided logger.
In that case, we should as you're doing here, but then flush to the default logger and exit with an error.

File: app.go
Comment (shirchen) [reply app.go:479]: Is this for readability? I'm curious to learn about which way is better and why.

File: app.go
Comment (shirchen) [reply app.go:525]: Yup, good call. Tests didn't catch this since I constructed a spy logger.

File: app.go
Comment (shirchen) [reply app.go:265]: I see, is there harm in leaving this public so we can keep on using this in tests and elsewhere and isolate exporting fxlog.Logger?

File: app.go
Comment (shirchen) [reply app.go:512]: _I'm not sure but I think this check should be after opt.apply but before app.provide_

So WithLogger is going to be part of Provide, so I think we need this after app.provide as `fxlogConstructor` is populated by WithLogger opt.apply but it'll need `zap.Logger` initialized by something like `zapfx`.

Good call on trying to simplify logBuffer, let me see how I can make it work.

File: app.go
Comment (abhinav) [reply app.go:265]: The reason I'd want to unexport is because it's currently unusable. As a user of Fx, there's no valid way for me to call this function right now.

That said, if you think we won't release this change until we also finish exporting fxlog.Logger, I'm okay with keeping it exported in this PR.

File: app.go
Comment (abhinav) [reply app.go:512]: If we do this after the `app.provides`, then we've already called `container.Provide` and logged a bunch before we've even considered whether there was a user-provided logger, right?

It's an option, but it's a "more important" option than provides. Like, we run Provides first, and Invokes later. We should examine WithLogger *before* we start logging things the same way we consider the current fx.Printer before anything else.

To be clear: we won't construct the logger until after app.provides, but we need to know whether we're buffering or using the default logger before we do app.provides. Please read the code samples above again.

File: app.go
Comment (shirchen) [reply app.go:265]: Yea, I think there isn't a point in releasing until this API is usable, so exporting fxlog.Logger should be a blocker for next release.

File: app.go
Comment (abhinav) [inline app.go:468]: We don't need to initialize this field here because it's always overwritten (L481-493).
```suggestion
```

File: app.go
Comment (abhinav) [inline app.go:500]: If a custom logger was provided, and this fails (because there was an error with one of the provides, like `fx.Provide(NewFoo, NewFoo)`), we won't print anything and fail silently.

We need to flush the buffer if there was a buffer, so maybe something like this:

```suggestion
	if app.err != nil {
		connectLogger(fxlog.DefaultLogger(os.Stderr))
		return app
```

File: app.go
Comment (abhinav) [inline app.go:514]: This will never be true unless there's a massive bug up above, in which case we should just panic. I would drop this else block and remove the `if` on connectLogger. Part of the contract of this code here is that if a custom constructor was provided, connectLogger was set.

File: app.go
Comment (abhinav) [inline app.go:510]: nit: When there are just two errors, multierr.Append is more efficient.
```suggestion
			app.err = multierr.Append(app.err,
				fmt.Errorf("could not construct customer logger via fx.WithLogger: %w", err))
```

File: app.go
Comment (abhinav) [inline app.go:513]:
```suggestion
				app.err = multierr.Append(app.err, err)
```

File: app.go
Comment (abhinav) [inline app.go:507]: This if block is massive. Can we pull it into a method responsible for building the logger and nothing else?

```
if app.fxlogConstructor != nil {
  log, err := app.buildCustomLogger()  // better name TBD
  if err != nil {
    // failed to build custom logger. Flush to stderr.
    connectLogger(fxlog.DefaultLogger(os.Stderr))
    app.err = err
    return app
  }
  connectLogger(log)
}
```

File: app.go
Comment (abhinav) [inline app.go:531]: We shouldn't manipulate app.err inside the Invoke. If we got to the Invoke, we know app.err is blank, so we can just return the error.

```
err := Invoke(func(log fxlog.Logger) error {
  app.log = log
  return connectLogger(app.log)
})

if err != nil {
  app.err = err
  // ...
}
```

If there are any errors in invoking (because the logger could not be constructed), or because `connectLogger` failed, that'll be surfaced in `err`.

File: log.go
Comment (abhinav) [inline log.go:49]: We should remove this check. This check is the sole reason that Connect has to return an error, and that's making all the logic above significantly more complicated. If we make `Connect` non-error-able, the log-construction logic gets simplified signfiicantly.

We don't really need to handle the "no logger" case here. This isn't an API external users can call. We have control over it. We can assume that we'l always call it correctly, or make necessary checks in the callers.

File: log.go
Comment (abhinav) [inline log.go:54]: nil is a valid slice and costs no allocs

```suggestion
	l.entries = nil
```

File: app.go
Comment (shirchen) [reply app.go:514]: I have/had this check due to a nil deference linter, which points out that its possible to have connectLogger be nil here. I thought that we wanted to stay on the good side of it, but happy to remove this if you dont think this adds value.

File: app.go
Comment (shirchen) [reply app.go:500]: ugh, we already have a test for this case (line 513 in `app_test.go` but since it's using a spy, we didn't have a test about flushing out the actual message. i think testing this will be easier once #747 lands since i added a way to setup zap and see what exactly gets printed out. i'll address this comment after rebase.

File: app.go
Comment (abhinav) [reply app.go]: If we're trying to appease the linter here, I'm okay with a manual `panic` here because that's an irrecoverable "there;s a massive bug in Fx" sort of scenario.

File: app.go
Comment (abhinav) [reply app.go:311]: +1

File: app.go
Comment (abhinav) [inline app.go:458]: nit:

 ```suggestion
```

File: app.go
Comment (abhinav) [inline app.go:464]: nit:
```suggestion
```

File: app.go
Comment (abhinav) [inline app.go:461]: nit:
```suggestion

	err := app.container.Invoke(
```

File: app.go
Comment (abhinav) [inline app.go:472]: nit: can shorten
```suggestion
	return err
```

File: app.go
Comment (abhinav) [inline app.go:499]: outdated comment
```suggestion
		connectLogger = buffer.Connect // func(fxlog.Logger)
```

File: app.go
Comment (abhinav) [inline app.go:519]: nit: the other comments are helpful but this is a bit extraneous
```suggestion
```

File: app.go
Comment (abhinav) [inline app.go:521]: nit: can inline
```suggestion
		if err := app.constructCustomLogger(connectLogger); err != nil {
```

File: app_test.go
Comment (abhinav) [inline app_test.go:282]: This will write to stdout. Maybe we should use a zaptest.NewLogger here so it only writes to the testing.T.

```suggestion
			Supply(zaptest.NewLogger(t)),
```

File: app_test.go
Comment (abhinav) [inline app_test.go:293]: This and the development logger are the same tests as the one above, with the difference being whether they'll write to stdout or stderr. In the interest of keeping stdout/stderr as silent as possible, I would suggest omitting these two since the one above will already cover the "instantiate a logger with no dependencies" case.

File: app_test.go
Comment (abhinav) [inline app_test.go:331]: Can we add a couple more test cases? In particular, we should test a logger with a dependency on other provided objects. This matches what we'll be using these APIs for. We should have these cases:

- Success
- Dependency failed to build because of an error
- Logger failed to build because of an error

We can test the output of the success case pretty easily: have the logger write to a buffer.
However, to test the output of the failure cases that write to stderr, we'll have to temporarily hijack os.Stderr by replacing it temporarily with an `os.File` for a temporary file (`ioutil.TempFile(t.TempDir(), "stderr")`?) and asserting its output.

File: fxevent/event.go
Comment (abhinav) [inline fxevent/event.go:45]: Forgot to add these to the `event()` test?

File: fxevent/event.go
Comment (abhinav) [inline fxevent/event.go:106]: CustomLoggerError to match CustomLogger?

File: fxevent/zap.go
Comment (abhinav) [inline fxevent/zap.go:77]: nit:
```suggestion
		l.Logger.Info("installing custom fxevent.Logger", zap.String("function", fxreflect.FuncName(e.Function)))
```

File: fxevent/zap.go
Comment (abhinav) [inline fxevent/zap.go:75]: nit: why we were unable to build the logger should be in the error message; the extra string can be skipped
```suggestion
		l.Logger.Error("error constructing logger", zap.Error(e.Err))
```

**Review by abhinav — COMMENTED** [review]
Minus need for test cases for logger dependencies LGTM!

File: app.go
Comment (shirchen) [reply app.go]: sg, though adding test coverage will be impossible so i changed my mind: its easier to ignore the lint warning.

File: app.go
Comment (shirchen) [reply app.go:458]: Here, I try and follow https://github.com/ssgreg/nlreturn. Do we disagree that this is a valid lint warning?

Comment (abhinav) [reply app.go:458]: That linter seems pretty arbitrary, TBH. I feel like whether a newline should be added in the function body or not is a matter of "paragraphs" of code with a shared purpose, and not something you can always lint for.

This return here is connected to the two statements above it, in that all three together are how we're choosing to handle that error. Like, how we handle that error is `(record, connect, return)`, not `(record, connect)` and an independent `return`. So to me the `return` is part of the paragraph containing the other two statements.

Comment (shirchen) [reply app.go:458]: ok, sounds good. i'll disable it from list that i use.

**Review by abhinav — APPROVED** [review]
LGTM; please review the additions I've made as well.

CC @r-hang

**Review by sywhang — APPROVED** [review]
LGTM!

**Pull request merged**
